### PR TITLE
Ticket647 keithley 2700 test files

### PR DESCRIPTION
**Pull request by esouthren**
Tests for Keithley 2700 - Part of HIFI Cryomagnet work

### CODE REVIEW

**Comment by AdrianPotter** [comment]
https://github.com/ISISComputingGroup/IBEX/issues/647

File: tests/keithley_2700.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    def setUp(self):
+        self._lewis, self._ioc = get_running_lewis_and_ioc("keithley_2700", DEVICE_PREFIX)
+        self.ca = ChannelAccess(default_timeout=30, device_prefix=DEVICE_PREFIX)
+        self.ca.wait_for("IDN")
+
+    def test_GIVEN_scan_state_set_WHEN_read_THEN_scan_state_is_as_expected(self):
+        sample_data = [0, 1]
+        expected_state = ["INT", "NONE"]
+        for set_value, expected_value in zip(sample_data, expected_state):
```
Comment (AdrianPotter) [inline tests/keithley_2700.py]: I'm not a fan of zipping. It means you have to be certain your data structures are in the right order. If the test is ever modified, there's potential for confusion. If your data is logically connected, it should be obvious (if possible) at point of definition:

```
sample_data = {0: "INIT", 1: "NONE"}
for set_value, expected_value in sample_data.items():
    ...
```

Please apply throughout

File: tests/keithley_2700.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    Tests for the Keithley2700.
+    """
+
+    def setUp(self):
+        self._lewis, self._ioc = get_running_lewis_and_ioc("keithley_2700", DEVICE_PREFIX)
+        self.ca = ChannelAccess(default_timeout=30, device_prefix=DEVICE_PREFIX)
+        self.ca.wait_for("IDN")
+
+    def test_GIVEN_scan_state_set_WHEN_read_THEN_scan_state_is_as_expected(self):
```
Comment (AdrianPotter) [inline tests/keithley_2700.py]: Two points here

- I think your `GIVEN` is a `WHEN`. The test should describe behaviour, not functions. `WHEN_I_set_the_scan_state_THEN_the_readback_matches_the_set_state`. If there's no special context for the test, then the `GIVEN` can be omitted
- One of my pet peeves in tests, ask @Tom-Willemsen, is "as expected" or "is correct". It doesn't add any worthwhile information to the test name. A test name should be self-describing in terms of expected behaviour. "as expected" simply confers that this is a test with an expected result but doesn't confer anything about the system under test.

Please apply throughout.

File: tests/keithley_2700.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        sample_data = [0, 1, 2, 3, 4]
+        expected_channel = ["IMM", "TIM", "MAN", "BUS", "EXT"]
+        for set_value, expected_value in zip(sample_data, expected_channel):
+            self.ca.assert_setting_setpoint_sets_readback(set_value, "CONTROLSOURCE", "CONTROLSOURCE:SP",
+                                                          expected_value)
+
+    def test_GIVEN_buffer_size_set_WHEN_read_THEN_buffer_size_is_as_expected_AND_within_range(self):
+        expected_buffer_size = [5500, 0, 2, 70000]
+        alarm_state = [ChannelAccess.ALARM_NONE, ChannelAccess.ALARM_MAJOR, ChannelAccess.ALARM_NONE,
+                       ChannelAccess.ALARM_MAJOR]
```
Comment (AdrianPotter) [inline tests/keithley_2700.py]: You appear to be testing the positive and negative cases alongside here, but that isn't indicated in the name of the test case. Please keep tests atomic.

File: tests/keithley_2700.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        expected_channel = ["SENS", "CALC", "NONE"]
+        for set_value, expected_value in zip(sample_data, expected_channel):
+            self.ca.assert_setting_setpoint_sets_readback(set_value, "BUFFER:SOURCE", "BUFFER:SOURCE:SP",
+                                                          expected_value)
+
+    def test_GIVEN_init_state_set_WHEN_read_THEN_init_state_is_as_expected(self):
+        sample_data = [0, 1]
+        expected_state = ["ON", "OFF"]
+        for set_value, expected_value in zip(sample_data, expected_state):
+            self.ca.assert_setting_setpoint_sets_readback(set_value, "INITMODE", "INITMODE:SP", expected_value)
```
Comment (AdrianPotter) [inline tests/keithley_2700.py]: Does this method not assume the structure of the set point (i.e. `<prefix>:SP`) PV if unspecified? If so, please remove the superfluous argument. Apply throughout

File: tests/keithley_2700.py (hunk 3 above)
Comment (esouthren) [reply tests/keithley_2700.py]: Seperated tests

File: tests/keithley_2700.py (hunk 2 above)
Comment (esouthren) [reply tests/keithley_2700.py]: Changed test titles

File: tests/keithley_2700.py (hunk 4 above)
Comment (esouthren) [reply tests/keithley_2700.py]: Removed superfluous arguments

File: tests/keithley_2700.py (hunk 1 above)
Comment (esouthren) [reply tests/keithley_2700.py]: Removed zips

File: tests/keithley_2700.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
+import unittest
+from utils.channel_access import ChannelAccess
+from utils.ioc_launcher import get_default_ioc_dir
+from utils.test_modes import TestModes
+from utils.testing import get_running_lewis_and_ioc, skip_if_recsim
+import time
```
[the remark this answers is not in the document]
Comment (esouthren) [reply tests/keithley_2700.py]: Removed

File: tests/keithley_2700.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    def test_WHEN_delay_state_set_THEN_delay_state_matches_the_set_state(self):
+        for enum_value, string_value in on_off_status.items():
+            self.ca.assert_setting_setpoint_sets_readback(enum_value, "DELAYMODE", expected_value=string_value)
+
+    def test_WHEN_source_set_THEN_source_matches_the_set_state(self):
+        sample_data = {0: "IMM", 1: "TIM", 2: "MAN", 3: "BUS", 4: "EXT"}
+        for enum_value, string_value in sample_data.items():
+            self.ca.assert_setting_setpoint_sets_readback(enum_value, "CONTROLSOURCE", expected_value=string_value)
+
+    def test_WHEN_buffer_size_set_THEN_buffer_size_matches_the_set_state_AND_alarm_is_major(self):
```
[the remark this answers is not in the document]
Comment (esouthren) [reply tests/keithley_2700.py]: Test passing confirmed

**Pull request merged**